### PR TITLE
Adding Support to GCP

### PR DESCRIPTION
**Pull request by jesayafn**
# Adding Support to GCP
Related issue: #30 

## Feature enhancement:
- New Attack Surface 
- New GCP Attack Technique:  **GCPEstablishAccessAsServiceAccount**
- Capability to save established GCP credential
- Capability to delete current/selected GCP credential

### CODE REVIEW

**Comment by jesayafn** [comment]
Hi @openrec0n, can you add "WIP" label on this PR?

**Comment by jesayafn** [comment]
Hi @openrec0n, you can review the PR

**Comment by openrec0n** [comment]
@jesayafn thanks for the PR. I am reviewing this

File: Halberd.py [hunk 1]
```diff
@@ -2179,6 +2217,62 @@ def manage_progress_display(execute_clicks, view_clicks, is_open):
         
     raise PreventUpdate
 
+'''C062 - Callback to generate GCP session options in AWS sessions dropdown'''
```
Comment (openrec0n) [inline Halberd.py:2220]:
```suggestion
'''C062 - Callback to generate GCP session options in GCP sessions dropdown'''
```

Comment (openrec0n) [inline Halberd.py:2220]:
```suggestion
'''C062 - Callback to generate GCP session options in GCP sessions dropdown'''
```

File: attack_techniques/gcp/gcp_establish_access_as_sa.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        super().__init__("Establish Access As Service Account", "Establishes access to Google Cloud Platform as service account", mitre_techniques)
+
+    def execute(self, **kwargs: Any) -> Tuple[ExecutionStatus, Dict[str, Any]]:
+        self.validate_parameters(kwargs)
+        try:
+            # input validation
+            raw_credential: str = kwargs['credential']
+            name: str = kwargs['name']
+            save_and_activate: bool = kwargs['save_and_activate']
+            set
```
Comment (openrec0n) [inline attack_techniques/gcp/gcp_establish_access_as_sa.py:31]:
```suggestion
```

File: attack_techniques/gcp/gcp_establish_access_as_sa.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            
+            # Input validationc
```
Comment (openrec0n) [inline attack_techniques/gcp/gcp_establish_access_as_sa.py:33]:
```suggestion
            # Input validation
```

File: attack_techniques/gcp/gcp_establish_access_as_sa.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                }
+            else :
+                caller_info_output["expired"] = False
+
+            if save_and_activate :
+                access_manager.save_credential()
+
+            return ExecutionStatus.SUCCESS, {
+                "value": caller_info_output,
+                "message": f"Successfully established access to target Azure tenant"
```
Comment (openrec0n) [inline attack_techniques/gcp/gcp_establish_access_as_sa.py:73]:
```suggestion
                "message": f"Successfully established access to target GCP tenant"
```

File: attack_techniques/gcp/gcp_establish_access_as_sa.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    "message": "Failed to establish access to GCP. The credential is expired"
+                }
+            else :
+                caller_info_output["expired"] = False
+
+            if save_and_activate :
+                access_manager.save_credential()
+
+            return ExecutionStatus.SUCCESS, {
+                "value": caller_info_output,
```
Comment (openrec0n) [inline attack_techniques/gcp/gcp_establish_access_as_sa.py:72]:
```suggestion
                "value": {
                    "credential_info": caller_info_output,
                    "access_status": {
                        "saved": save_and_activate,
                        "activated": save_and_activate,
                        "scopes": current_access.scopes,
                        "token_expiry": str(current_access.expiry) if current_access.expiry else None
                    }
                }
```

Comment (openrec0n) [reply attack_techniques/gcp/gcp_establish_access_as_sa.py:72]: Suggested change to return a more detailed output with access information

File: core/gcp/gcp_access.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+import json
+from core.Constants import GCP_CREDS_FILE
+
+class GCPAccess():
+    """GCP access manager"""
+    credential = None
+    
+
+
+
```
Comment (openrec0n) [inline core/gcp/gcp_access.py:17]:
```suggestion
    
```

File: core/gcp/gcp_access.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                self.encoded_credential=base64.b64encode(raw_credentials.encode())
+                # # Load the credentials using the   parsed JSON and provided scopes
+                # credentials = service_account.Credentials.from_service_account_info(
+                #     raw_credentials,
+                #     scopes=scopes
+                # )
+                
+                # # Initialize the parent class (service_account.Credentials) with extracted values
+                # super().__init__(credentials.signer, credentials.service_account_email, credentials._token_uri, scopes=scopes)
+
```
Comment (openrec0n) [inline core/gcp/gcp_access.py:43]: @jesayafn these comments can be cleared out

File: core/gcp/gcp_access.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
+import base64
+import os
+from typing import Union
```
Comment (openrec0n) [inline core/gcp/gcp_access.py:3]:
```suggestion
```

Comment (openrec0n) [reply core/gcp/gcp_access.py:3]: Suggested change to remove import since its not being used

**Review by openrec0n — CHANGES_REQUESTED** [review]
@jesayafn Please check the comments and implement the suggested changes.

File: Halberd.py [hunk 9]
```diff
@@ -235,6 +237,25 @@ def execute_technique_callback(n_clicks, tactic, t_id, values, bool_on, file_con
...
+        try:
+            current_access = None
+            if t_id == "GCPEstablishAccessAsServiceAccount":
+                manager = GCPAccess(raw_credentials=file_content[0],name=values[0])
+                current_access = manager.get_current_access().get("name")
+                current_access = manager.get_detailed_credential(name=current_access)
+            else:
+                manager = GCPAccess()
+                current_access = manager.get_current_access()
+            if current_access["credential"]["type"] == "service_account":
```
[the remark this answers is not in the document]
Comment (jesayafn) [reply Halberd.py:250]: image

`current_access` has a `type` field under `credential` dict.

**Review by openrec0n — APPROVED** [review]

**Pull request merged**